etcd: Fix several etcd related issues

tgraf opened the pull request:
See individual commits. The most important fix is in the last commit.

aanm commented on pkg/kvstore/etcd.go:
```diff
@@ -544,8 +544,21 @@ func (e *etcdClient) renewSession(ctx context.Context) error {
+	case <-e.firstSession:
+	// controller has stopped or etcd client is closing
+	case <-ctx.Done():
+		return nil
```
Shouldn't be `return ctx.Err()`?

tgraf replied:
Depends, the context is cancelled when the controller stops or when the etcd client is shut down. So `nil` is returned to not report a controller failure if the controller stopped.

tgraf replied:
Note that a later commit changes this to return the `err` after `firstSession` started returning the error observed.

aanm commented on pkg/kvstore/etcd.go:
```diff
@@ -544,8 +544,21 @@ func (e *etcdClient) renewSession(ctx context.Context) error {
+	case <-e.session.Done():
+	// controller has stopped or etcd client is closing
+	case <-ctx.Done():
+		return nil
```
Shouldn't be `return ctx.Err()`?

tgraf replied:
This is still returning `nil` after all changes unlike the previous case. The code is choosing to not return an error to not trigger a retry when the controller stops or the etcd client shuts down.

aanm commented on pkg/kvstore/etcd.go:
```diff
@@ -544,8 +544,21 @@ func (e *etcdClient) renewSession(ctx context.Context) error {
+
+	select {
+	// session has ended
+	case <-e.session.Done():
```
shouldn't be `<-e.lockSession.Done()`?

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -351,7 +351,7 @@ func (c *CNPStatusUpdateContext) updateViaKVStore(ctx context.Context, cnp *type
 
 	if err := <-kvstore.Client().Connected(ctx); err != nil {
-		return nil
+		return fmt.Errorf("kvstors is unavailable: %s", err)
```
`s/%s/%w`

qmonnet replied:
And `s/kvstors/kvstore/`

tgraf committed:
k8s: Exit CNP status update controller if kvstore is unavailable

Commit 19ad311f5c incorrectly changed the return value to nil instead of
returning an error.

Fixes: 19ad311f5c1 ("kvstore: Fix Watch() to return when client is closed or context is cancelled")
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
etcd: Ensure that lock session renewal controller exit

Same as ba9031bb45a ("etcd: Ensure that session renewal controller exit") but
for the lock session.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
etcd: Ensure that firstSession is closed

The firstSession channel has been used to allow blocking until the initial
session has been established. However, because callers have not been handling
errors, it was not possible to close the channel if the etcd client was shut
down before the session was ever established and thus the channel was leaked.

This was even more problematic for etcd operations performed without a context
with a timeout as most operations currently block on firstSession or context.

Consolidate the errChan and firstSession as they effectively served the same
purpose. The only slight change is that the firstSession channel is closed
*after* the etcd version has been verified which is the right behavior anyway.

Having the error condition returned via the firstSession channel allows to
return it to owning controllers as well when renewing sessions for better
visibility.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
etcd: Print info message when connection is established

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
etcd: Lock e.session when renewing session

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
etcd: Report reason when lock acquisition in status check fails

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
etcd: Fix incorrect context usage in session renewal

The context passed into NewSession() was supposed to enforce a timeout on the
NewSession() operation which is triggering a Grant() and KeepAlive()
instruction. However, the context passed into NewSession() will also be
associated with the resulting lease. As per etcd documentation, this will
result in:

> If the context is canceled before Close() completes, the session's lease will
> be abandoned and left to expire instead of being revoked.

Because of this, any session renewal triggering a new session would create a
session that is immediately closed again due to the context passed into
NewSession being cancelled when the controller run ends successfully. This
resulted in any renewed session to have an effective lifetime of 10
milliseconds before requiring renewal again.

Fixes: #12619
Fixes: 8524fca879b ("kvstore: Add session renew backoff")

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on pkg/k8s/cnp.go:
```diff
@@ -351,7 +351,7 @@ func (c *CNPStatusUpdateContext) updateViaKVStore(ctx context.Context, cnp *type
 
 	if err := <-kvstore.Client().Connected(ctx); err != nil {
-		return nil
+		return fmt.Errorf("kvstore is unavailable: %w", err)
```
nit:

```suggestion
		return errors.Wrap(err, "kvstore is unavailable")
```

tgraf replied:
I think we are sticking to `fmt.Errorf()`. @tklauser has recently swapped instances of `errors.Wrap()` over to `fmt.Errorf()` again.

tklauser replied:
FWIW, this was #12603. The idea was to use the functionality available in the Go standard library (since Go 1.13) rather than an external package for error wrapping.

joestringer commented on pkg/kvstore/etcd.go:
```diff
@@ -652,7 +648,7 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
 		session:              &s,
 		lockSession:          &ls,
-		firstSession:         firstSession,
+		firstSession:         errChan,
```
`errChan` or `errorChan`? :thinking:

tgraf replied:
`errChan` is closed after blocking on `errorChan` so the `firstSession` won't be closed if `errorChan` receives any errors. I think we can eventually consolidate `errChan` and `errorChan` eventually, just didn't want to overload this PR too much.

joestringer commented on pkg/kvstore/etcd.go:
```diff
@@ -783,11 +778,22 @@ func (e *etcdClient) checkMinVersion(ctx context.Context) error {
+		}
 	case <-ctx.Done():
-		return nil, fmt.Errorf("lock cancelled via context: %s", ctx.Err())
+		return fmt.Errorf("interrupt while waiting for initial session to be established: %s", ctx.Err())
```
nit: `errors.Wrap()`

aanm replied:
nit: `%w`

joestringer commented on pkg/contexthelpers/context_test.go:
```diff
@@ -0,0 +1,58 @@
+	select {
+	case <-ctx.Done():
+	case <-time.After(time.Second):
+		c.Errorf("no timeout despite reporting success")
```
nit: Isn't this more like "conditional timeout was not triggered"?

I found this confusing as 'no timeout' could apply to this timeout from `time.After()` or the timeout configured in the NewConditionalTimeoutContext() call above, and also I don't understadn "despite reporting success" since this path through the test case doesn't send a success to the channel.

joestringer commented on pkg/kvstore/etcd.go:
```diff
@@ -568,20 +588,35 @@ func (e *etcdClient) renewLockSession(ctx context.Context) error {
+		e.lockSessionCancel = nil
+	}
+
+	// Create a context representing the lifetime of the session. It will
```
nit:

```suggestion
	// Create a context representing the lifetime of the lock session. It will
```

joestringer reviewed (changes requested):
One question below on `errChan` vs. `errorChan` that needs to be checked.

The rest are misc. nits. Not sure whether we've collectively decided on `errors.Wrap()` but I know some parts of the code have started using it already.

christarazi commented on pkg/kvstore/etcd.go:
```diff
@@ -249,8 +249,10 @@ func Hint(err error) error {
 	// firstSession is a channel that will be closed once the first session
-	// is set up in the etcd Client.
-	firstSession chan struct{}
+	// is set up in the etcd client. If an error occurred and the initial
```
Typo in commit msg:

`established an thus the channel was leaked.`   

`established and thus the channel was leaked.`

christarazi commented on pkg/contexthelpers/context_test.go:
```diff
@@ -0,0 +1,58 @@
+	c.Assert(cancel, check.Not(check.IsNil))
+	c.Assert(ch, check.Not(check.IsNil))
+
+	select {
```
I think it would be helpful to describe what this test is doing and what we expect. This way we can document the behavior of this helper for our future selves if we ever have another use-case for it.

christarazi commented on pkg/kvstore/etcd.go:
```diff
@@ -520,21 +525,36 @@ func (e *etcdClient) renewSession(ctx context.Context) error {
+	// Create a context representing the lifetime of the session. It will
+	// timeout if the session creation does not succeed in time and then
+	// persists until:
+	//  - the parent context is cancelled due to the etcd client closing or
```
Nit to be consistent with the rest of the comment: 

```suggestion
	//  - The parent context is cancelled due to the etcd client closing or
```

christarazi commented on pkg/kvstore/etcd.go:
```diff
@@ -520,21 +525,36 @@ func (e *etcdClient) renewSession(ctx context.Context) error {
+	// persists until:
+	//  - the parent context is cancelled due to the etcd client closing or
+	//    the controller being shut down
+	//  - The above call to sessionCancel() cancels the session due to the
```
> It will timeout if the session creation does not succeed in time and then persists until:

Are the conditions that follow the above statement, ANDed or ORed? In other words, does the context persist when the parent context is cancelled (AND|OR) when the above call to `sessionCancel()`?

christarazi commented on pkg/kvstore/etcd.go:
```diff
@@ -568,20 +588,35 @@ func (e *etcdClient) renewLockSession(ctx context.Context) error {
+	// timeout if the session creation does not succeed in time and then
+	// persists until:
+	//  - the parent context is cancelled due to the etcd client closing or
+	//    the controller being shut down
```
Same here with the capitalization and clarification of the conditions.

christarazi reviewed (approved):
LGTM at the surface; not too familiar with this code so I'll rely on others' reviews on that. Reviewed changes from a semantic perspective mostly.

joestringer reviewed (approved):
Questions asked & answered. Minor nits can be addressed now or as followup.

aanm commented on pkg/kvstore/etcd.go:
```diff
@@ -381,25 +381,24 @@ func (e *etcdClient) closeLockSession(leaseID client.LeaseID) {
 			select {
 			case <-e.client.Ctx().Done():
-				initLockSucceeded <- false
+				initLockSucceeded <- fmt.Errorf("client context ended")
```
similar to what is done on line 395, why not `fmt.Errorf("client context ended: %w", e.client.Ctx().Err())`?

aanm reviewed (approved):
nice finding of these annoying bugs. 2 small nits
